### PR TITLE
Implement bloom/api/fonts/names and bloom/api/fonts/metadata (BL-10387)

### PR DESCRIPTION
**Pull request by StephenMcConnel**
fonts/metadata is implemented only for Windows in this commit.
fonts/names is implemented for both Windows and Linux.

---
This change is [Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812)

### CODE REVIEW

**Review by andrew-polk — CHANGES_REQUESTED** [review]
A few minor comments. But this is one @hatton will probably want to take a peek at.

> Reviewed 14 of 14 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-:-MnSWxLZ-70aOOU2WFtH:bqz7naw)* status: all files reviewed, 2 unresolved discussions (waiting on @StephenMcConnel)

___
*[src/BloomExe/FontProcessing/FontMetadata.cs, line 73 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnSWLOF4fCOVQgMezj9:-MnSWLOF4fCOVQgMezjA:b-bvnduw) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/0ef4a3a95def76f7bfbb803a966916d6dfeb7584/src/BloomExe/FontProcessing/FontMetadata.cs#L73)):*
> ```
> version = gtf.VersionStrings[english];
> ```
This line will throw if we caught an exception above.
___
*[src/BloomExe/FontProcessing/FontsApi.cs, line 76 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnSVwmHCgReet9l_RdI:-MnSVwmI8J2py6tgW5XA:blykax6) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/0ef4a3a95def76f7bfbb803a966916d6dfeb7584/src/BloomExe/FontProcessing/FontsApi.cs#L76)):*
> ```Smalltalk
> 
> 		/// <summary>
> 		/// Return a list of FontMetadata objects for all fonts that we currently know about.
> ```

Worth being more explicit in the comment for these that we may not have populated the data yet?
(Or am I misunderstanding?)

**Review by StephenMcConnel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812)* status: all files reviewed, 2 unresolved discussions (waiting on @andrew-polk)

___
*[src/BloomExe/FontProcessing/FontMetadata.cs, line 73 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnSWLOF4fCOVQgMezj9:-MnWq9lABQcK9YzgUNEm:bmm3n54) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/0ef4a3a95def76f7bfbb803a966916d6dfeb7584/src/BloomExe/FontProcessing/FontMetadata.cs#L73)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> version = gtf.VersionStrings[english];
> ```
This line will throw if we caught an exception above.
</blockquote></details>

Good catch.  I noticed this and moved everything inside the try block.
___
*[src/BloomExe/FontProcessing/FontsApi.cs, line 76 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnSVwmHCgReet9l_RdI:-MnWrIc38xtInYO9o6PP:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/0ef4a3a95def76f7bfbb803a966916d6dfeb7584/src/BloomExe/FontProcessing/FontsApi.cs#L76)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Worth being more explicit in the comment for these that we may not have populated the data yet?
(Or am I misunderstanding?)
</blockquote></details>

Done.

**Review by andrew-polk — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812)* status: all files reviewed, 1 unresolved discussion (waiting on @StephenMcConnel)

___
*[src/BloomExe/FontProcessing/FontMetadata.cs, line 88 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnY-IhL9-E_ZvnzEjsq:-MnY-IhL9-E_ZvnzEjsr:b-h1vhw9) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/70a8b0c8bb2f63ffc09228750b5d9b40b814515f/src/BloomExe/FontProcessing/FontMetadata.cs#L88)):*
> ```
> otfinfo
> ```
We can assume Linux users will have this installed?
In other words, we don't need to add this package to the debian config stuff?

**Review by StephenMcConnel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812)* status: 14 of 15 files reviewed, 1 unresolved discussion (waiting on @andrew-polk)

___
*[src/BloomExe/FontProcessing/FontMetadata.cs, line 88 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnY-IhL9-E_ZvnzEjsq:-MnaA4P6ANLblzXR-eBh:b-wq5sy5) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/70a8b0c8bb2f63ffc09228750b5d9b40b814515f/src/BloomExe/FontProcessing/FontMetadata.cs#L88)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> otfinfo
> ```
We can assume Linux users will have this installed?
In other words, we don't need to add this package to the debian config stuff?

</blockquote></details>

You're right.  I had forgotten to add the package to the debian control file when I installed it last week.  I thought of it last night...

**Review by andrew-polk — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @StephenMcConnel)

File: src/BloomExe/FontProcessing/FontMetadata.cs [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+		public string designer { get; private set; }
+		public string designerURL { get; private set; }
+		public string trademark { get; private set; }
+		public string determinedSuitability { get; private set; }
+		public string determinedSuitabilityNotes { get; private set; }
+
+		public FontMetadata(string name, FontGroup group)
+		{
+			this.name = name;
+#if __MonoCS__
```
Comment (hatton) [inline src/BloomExe/FontProcessing/FontMetadata.cs:33]: Worth a comment, like "GlyphTypeFace is better (why?) so we're using that for .net, but for mono we have to use SharpFont.

File: src/BloomExe/FontProcessing/FontMetadata.cs [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+						determinedSuitabilityNotes = "Open Font License";
+					else if (license.StartsWith("Licensed under the Apache License"))
+						determinedSuitabilityNotes = "Apache";
+					else if (license.Contains("GNU LGPL") || license.Contains("GNU Lesser General Public License"))
+						determinedSuitabilityNotes = "GNU LGPL";
+					else
+						determinedSuitabilityNotes = "GNU GPL";
+					return;
+				}
+				if (license.Replace('\n',' ').Contains("Permission is hereby granted, free of charge, to any person obtaining a copy of the fonts accompanying this license (\"Fonts\") and associated documentation files (the \"Font Software\"), to reproduce and distribute the Font Software, including without limitation the rights to use, copy, merge, publish, distribute, and/or sell copies of the Font Software, and to permit persons to whom the Font Software is furnished to do so") &&
```
Comment (hatton) [inline src/BloomExe/FontProcessing/FontMetadata.cs:178]: Feels a bit brave to require the exact wording? Would "free of charge" be enough?

File: src/BloomExe/FontProcessing/FontMetadata.cs [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+			if (licenseURL == "http://dejavu-fonts.org/wiki/License")
+			{
+				determinedSuitability = "ok";
+				determinedSuitabilityNotes = "Bitstream free license";
+				return;
+			}
+			if (!String.IsNullOrEmpty(copyright))
+			{
+				// some people put license information in the copyright string.
+				if (copyright.Contains("This font is free; you can redistribute it and/or modify it under the terms of the Artistic License."))
```
Comment (hatton) [inline src/BloomExe/FontProcessing/FontMetadata.cs:195]: Would "Artistic License" be enough?

File: src/BloomExe/FontProcessing/FontMetadata.cs [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+				designer = gtf.DesignerNames[english];
+				designerURL = gtf.DesignerUrls[english];
+				license = gtf.LicenseDescriptions[english];
+				manufacturer = gtf.ManufacturerNames[english];
+				manufacturerURL = gtf.VendorUrls[english];
+				trademark = gtf.Trademarks[english];
+			}
+			catch (Exception)
+			{
+				// file is somehow corrupt or not really a font file? Just ignore it.
```
Comment (hatton) [inline src/BloomExe/FontProcessing/FontMetadata.cs:153]: Seems like at this point, we should be labelling the font in such a way as to indicate that we ran into an Error? E.g. can we set `determinedSuitabilityNotes`?

File: src/BloomExe/FontProcessing/FontMetadata.cs [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+				var standardOutput = process.StandardOutput.ReadToEnd();
+				var standardError = process.StandardError.ReadToEnd();
+				if (process.ExitCode == 0 && standardError.Length == 0)
+					ParseOtfinfoOutput(standardOutput);
+				else
+					Console.WriteLine("oftinfo -i \"{0}\" returned {1}.  Standard Error =\n{2}", group.Normal, process.ExitCode, standardError);
+			}
+			catch (Exception e)
+			{
+				Console.WriteLine("otfinfo -i \"{0}\" threw an exception: {1}", group.Normal, e);
```
Comment (hatton) [inline src/BloomExe/FontProcessing/FontMetadata.cs:107]: See note on the Windows scenario `catch` block

File: src/BloomExe/FontProcessing/FontsApi.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+		}
+
+		private static IEnumerable<FontMetadata> GetFontMetadataSortedByName()
+		{
+			var list = new List<FontMetadata>(_fontNameToMetadata.Values);
+			list.Sort((a,b) => a.name.CompareTo(b.name));
+			return list;
+		}
+
+		/// <summary>
```
Comment (hatton) [inline src/BloomExe/FontProcessing/FontsApi.cs:79]: Normally we put Public stuff at the top?

File: src/BloomExe/FontProcessing/FontsApi.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+			var list = new List<string>(Browser.NamesOfFontsThatBrowserCanRender());
+			list.Sort();
+			return list;
+		}
+
+		/// <summary>
+		/// Return a list of FontMetadata objects for all fonts on the system.  This will be SLOW the
+		/// first time it is called, but caches the result so that it will be fast in any later calls.
+		/// </summary>
+		public static IEnumerable<FontMetadata> GetAllFontMetadata()
```
Comment (hatton) [inline src/BloomExe/FontProcessing/FontsApi.cs:48]: Is the return just for test purposes? The normal call should be a background one that just loads up the static results? Would be helpful to say that.

File: src/BloomExe/FontProcessing/FontsApi.cs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+		public static IEnumerable<string> SortedListOfFontNames()
+		{
+			var list = new List<string>(Browser.NamesOfFontsThatBrowserCanRender());
+			list.Sort();
+			return list;
+		}
+
+		/// <summary>
+		/// Return a list of FontMetadata objects for all fonts on the system.  This will be SLOW the
```
Comment (hatton) [inline src/BloomExe/FontProcessing/FontsApi.cs:45]: I usually don't like comments to say much about who uses them, but in this case I think it's worth saying, since you talk about it being slow: This is called by program.cs on Bloom startup. If it is not done before the first API call comes in... (what happens?)

File: src/BloomExe/FontProcessing/FontsApi.cs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+		/// first time it is called, but caches the result so that it will be fast in any later calls.
+		/// </summary>
+		public static IEnumerable<FontMetadata> GetAllFontMetadata()
+		{
+			lock(_fontMetadataLock)
```
Comment (hatton) [inline src/BloomExe/FontProcessing/FontsApi.cs:50]: I tripped over this a bit.
First, why not just lock on `_fontNameToMetadata`?
Second, this is only called once, so what is the point? I think the point should be that any process reading this value should also lock on it.
Third, could the lock be limitted to the line that actually accesses `_fontNameToMetadata` ?

File: src/BloomExe/FontProcessing/FontsApi.cs [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+				var list = GetFontMetadataSortedByName();
+				var ending = DateTime.Now;
+				System.Diagnostics.Debug.WriteLine($"DEBUG: Collecting font metadata took {ending - starting}");
+				return list;
+			}
+		}
+
+		private static IEnumerable<FontMetadata> GetFontMetadataSortedByName()
+		{
+			var list = new List<FontMetadata>(_fontNameToMetadata.Values);
```
Comment (hatton) [inline src/BloomExe/FontProcessing/FontsApi.cs:74]: As far as I can tell, `List` is not thread safe, *and* you're not locking it before writing or reading.

**Review by StephenMcConnel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812)* status: 6 of 15 files reviewed, 10 unresolved discussions (waiting on @andrew-polk and @hatton)

___
*[src/BloomExe/FontProcessing/FontMetadata.cs, line 33 at r3](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnRVSV8vQJ-Nx1-jYBM-r3-33:-MqWWP0qEkpuebY1-LPy:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f065a64b3af3425991c1408b4b5e6a1210919/src/BloomExe/FontProcessing/FontMetadata.cs#L33)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

Worth a comment, like "GlyphTypeFace is simpler so we're using that for .net, but for mono we have to use SharpFont + running a otfinfo as a command line process on each font." We get otfinfo as part of the lcdf-typetools package that is specified in the debian/control file.
</blockquote></details>

Done.
___
*[src/BloomExe/FontProcessing/FontMetadata.cs, line 107 at r3](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnRVSV8vQJ-Nx1-jYBM-r3-107:-Mq_bd4i1t0VJO-x_o9i:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f065a64b3af3425991c1408b4b5e6a1210919/src/BloomExe/FontProcessing/FontMetadata.cs#L107)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

See note on the Windows scenario `catch` block
</blockquote></details>

Done.
___
*[src/BloomExe/FontProcessing/FontMetadata.cs, line 153 at r3](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnRVSV8vQJ-Nx1-jYBM-r3-153:-MqWcuK45AkU_ar-iOKM:b-ssthh9) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f065a64b3af3425991c1408b4b5e6a1210919/src/BloomExe/FontProcessing/FontMetadata.cs#L153)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

Seems like at this point, we should be labelling the font in such a way as to indicate that we ran into an Error? E.g. can we set `determinedSuitabilityNotes`?
</blockquote></details>

Done.  I actually set determinedSuitability and fsType as well and immediately return in the catch blocks.
___
*[src/BloomExe/FontProcessing/FontMetadata.cs, line 178 at r3](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnRVSV8vQJ-Nx1-jYBM-r3-178:-Mq_fb4PClLkrGh6PpD5:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f065a64b3af3425991c1408b4b5e6a1210919/src/BloomExe/FontProcessing/FontMetadata.cs#L178)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

Feels a bit fragile to require the exact wording? Would "free of charge" (combined with "Bistream" be enough?
</blockquote></details>

Done.
___
*[src/BloomExe/FontProcessing/FontMetadata.cs, line 195 at r3](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnRVSV8vQJ-Nx1-jYBM-r3-195:-Mq_oVWyD7ulg2SY4U5M:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f065a64b3af3425991c1408b4b5e6a1210919/src/BloomExe/FontProcessing/FontMetadata.cs#L195)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

Would "Artistic License" be enough?
</blockquote></details>

Done.
___
*[src/BloomExe/FontProcessing/FontsApi.cs, line 45 at r3](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnRVSV8vQJ-Nx1-jYBN-r3-45:-Mq_s3JdCAV4H7VH7dZo:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f065a64b3af3425991c1408b4b5e6a1210919/src/BloomExe/FontProcessing/FontsApi.cs#L45)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

I usually don't like comments to say much about who uses them, but in this case I think it's worth saying, since you talk about it being slow: This is called by program.cs on Bloom startup. If it is not done before the first API call comes in... (what happens?)
</blockquote></details>

Done.
___
*[src/BloomExe/FontProcessing/FontsApi.cs, line 48 at r3](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnRVSV8vQJ-Nx1-jYBN-r3-48:-Mq_s1wN8TajCxnDvXTQ:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f065a64b3af3425991c1408b4b5e6a1210919/src/BloomExe/FontProcessing/FontsApi.cs#L48)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

Is the return just for test purposes? The normal call should be a background one that just loads up the static results? Would be helpful to say that.
</blockquote></details>

Done.
___
*[src/BloomExe/FontProcessing/FontsApi.cs, line 50 at r3](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnRVSV8vQJ-Nx1-jYBN-r3-50:-Mq_s-_T3yxD7em1PFvL:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f065a64b3af3425991c1408b4b5e6a1210919/src/BloomExe/FontProcessing/FontsApi.cs#L50)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

I tripped over this a bit.
First, why not just lock on `_fontNameToMetadata`?
Second, this is only called once, so what is the point? I think the point should be that any process reading this value should also lock on it.
Third, could the lock be limitted to the line that actually accesses `_fontNameToMetadata` ?
</blockquote></details>

Done.
___
*[src/BloomExe/FontProcessing/FontsApi.cs, line 74 at r3](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnRVSV8vQJ-Nx1-jYBN-r3-74:-MqaDSDxDDizQ8nL74Hm:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f065a64b3af3425991c1408b4b5e6a1210919/src/BloomExe/FontProcessing/FontsApi.cs#L74)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

As far as I can tell, `List` is not thread safe, *and* you're not locking it before writing or reading.
</blockquote></details>

Done.
___
*[src/BloomExe/FontProcessing/FontsApi.cs, line 79 at r3](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnRVSV8vQJ-Nx1-jYBN-r3-79:-MqaDQCU5KDPMMyrzLpd:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f065a64b3af3425991c1408b4b5e6a1210919/src/BloomExe/FontProcessing/FontsApi.cs#L79)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

Normally we put Public stuff at the top?
</blockquote></details>

Done.

**Review by hatton — APPROVED** [review]
> Reviewed 3 of 14 files at r1, 2 of 4 files at r2, 1 of 1 files at r3, 7 of 9 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812)* status: all files reviewed, 1 unresolved discussion (waiting on @StephenMcConnel)

___
*[src/BloomExe/FontProcessing/FontMetadata.cs, line 195 at r3](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnRVSV8vQJ-Nx1-jYBM-r3-195:-MqaJu8IDb-EgEQ3yiQY:b-odptca) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f065a64b3af3425991c1408b4b5e6a1210919/src/BloomExe/FontProcessing/FontMetadata.cs#L195)):*
<details><summary><i>Previously, StephenMcConnel (Steve McConnel) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

If you happen to do another round... could remove that trailing period. A font might not have it at the end of the sentence.

**Review by StephenMcConnel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812)* status: 14 of 15 files reviewed, 1 unresolved discussion (waiting on @hatton)

___
*[src/BloomExe/FontProcessing/FontMetadata.cs, line 195 at r3](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812#-MnRVSV8vQJ-Nx1-jYBM-r3-195:-MqaWubwAvcCKgAkqoTK:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f065a64b3af3425991c1408b4b5e6a1210919/src/BloomExe/FontProcessing/FontMetadata.cs#L195)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

If you happen to do another round... could remove that trailing period. A font might not have it at the end of the sentence.
</blockquote></details>

Done.

**Review by andrew-polk — APPROVED** [review]
> Reviewed 7 of 9 files at r4, 1 of 2 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812)* status: all files reviewed, 1 unresolved discussion (waiting on @hatton)

**Review by andrew-polk — APPROVED** [review]
> Dismissed @hatton from a discussion.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4812)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @StephenMcConnel)
